### ISSUE TITLE
clean up branches

### ISSUE THREAD

**Issue by davidak** [NixOS/nixpkgs#36755]
Some branches of this repository can probably get removed. Please check if the started work is worth to keep or already outdated.

- 0.5-stable (last activity in 2004) @edolstra 
- freebsd-losser (last activity in 2003) @edolstra 
- kmod-lib-modules (last activity in 2012) @urkud
- logistics (last activity in 2003) @edolstra 
- meteor (changes already in master https://github.com/NixOS/nixpkgs/blob/master/pkgs/servers/meteor/default.nix) @cstrahan
- mkMap ((last activity in 2014) @shlevy 
- nix-upgrade (PR closed, different solution merged in https://github.com/NixOS/nixpkgs/pull/24017) @domenkozar 
- purity-test (last activity in 2008) @edolstra 
- release-* (do we want to keep them forever?)
- revert-14975-pythonblas (PR merged https://github.com/NixOS/nixpkgs/pull/14978) @FRidh @grahamc why restored branch?
- revert-30286-patch-2 (PR merged https://github.com/NixOS/nixpkgs/pull/30385) @Profpatsch 
- revert-34178-fix/clang-python3 (PR merged https://github.com/NixOS/nixpkgs/pull/34193) @copumpkin
- security-updates (still needed) @nbp
- staging-17.03 (still needed) @edolstra 
- staging.patchShebangs (PR closed https://github.com/NixOS/nixpkgs/pull/19254) @domenkozar
- syscall-tracing (last activity in 2011) @edolstra 
- unity (no activity since 2012, meanwhile Canonical  announced they stop working on Unity) @aszlig 
- x86_64-darwin (last activity in 2009) @rbvermaa

**Comment by bjornfor** [comment]
IMHO we should remove everything but the release branches. (Development branches, stale or not, need not be pushed to the the main repository.)

**Comment by 7c6f434c** [comment]
@bjornfor active development branches with high impact or multiple collaborators might still make sense. For example, the staging branch stays. Of course, we are not talking about active branches here, so for the purposes of the current issue I agree.

**Comment by bjornfor** [comment]
Ok, staging branch is an exception.

**Comment by shlevy** [comment]
:+1: from me.

**Comment by nbp** [comment]
We should trash the security-updates branch. Nobody seems to be interested in doing any work on it, and it probably already bit-rot. Also, we are more likely to rewrite everything with something like SOS before getting any of the feature of the security-update branch.

**Comment by lheckemann** [comment]
Is staging-17.03 really still needed given that the release is EOL? I'm sceptical :)

**Comment by shlevy** [comment]
I'd actually like to go further than this. Git is distribtued anyway, if a given branch isn't on a special list of "this branch is persistent and never force-pushed" branches (which should be a subset of "this branch is protected on github" anyway), it should just be automatically deleted after N days of inactivity.

**Comment by bjornfor** [comment]
Or, since git is distributed, those branches need not be pushed to the main repo at all.  For instance, even the staging branches could go into a separate tree (e.g. nixos-staging), like we have for nixos-channels. Then the main repo is for stable, public consumption only.

**Comment by FRidh** [comment]
What problem exactly are we trying to solve here? Branches are occasionally created for development, typically when multiple members need to be able to push. Isn't that what the repository is for? Nixpkgs development? I agree we shouldn't just arbitrarily create new branches, but that hardly ever happens.

In any case, the majority of the listed branches can indeed be removed.

**Comment by shlevy** [comment]
@bjornfor But then we'd need some cleanup policy on that tree instead. I don't see a strong reason not to just use NixOS/nixpkgs as the home of non-permanent branches involving collaboration.

**Comment by bjornfor** [comment]
@shlevy: Right, a cleanup policy is still needed. My thoughts were just to limit the "branch noise" in the main repo.

Isn't it possible to work multiple people on PRs? Or if not, setup one nixos repo where people can push as many development branches as they like?

**Comment by shlevy** [comment]
You can work multiple people on PRs if they all have access toa  common tree or the person who owns it manually merges other work in. I think we should make "this is the branch cleanup policy for temporary branches on an official shared tree" a separate question from "this is where the official shared tree lives"

**Comment by grahamc** [comment]
Other than deleting merged branches, this doesn't seem like a very important thing. Old stale branches are old and stale, but who cares? Why care? the security updates branch while we may not merge it has valuable information and history there, perhaps useful for future work.



    release-* (do we want to keep them forever?)

yes probably

**Comment by davidak** [comment]
@grahamc when we can remove about 18 branches it would be easier to find the branch you want. when we don't clean up from time to time there will be 1000 branches some day. without a policy it's not clear what to remove and what to keep.

But yes, it's not an urgent problem to solve.

**Comment by 7c6f434c** [comment]
If you pull everything GitHub offers to pull, you are buried under PR branches more than anything. If you are goinf through a UI, they are sorted. Renaming the branches to something like `zz-stale/branch-name` might be somewhat useful.

**Comment by davidak** [comment]
@7c6f434c renaming branches on github would break push and pull for people who use the old name locally.

**Comment by grahamc** [comment]
True, however at the rate of 50 branches every 15 years maybe we've earned 1k branches by 2318 ;)

I agree with your point. We should definitely at least delete merged PR branches.

**Comment by shlevy** [comment]
I'd like to actually delete old stale branches because they litter git branch -a and the github branch list. What's the benefit of keeping them?

**Comment by edolstra** [comment]
Strong :-1:. Branches should not get removed unless they have been merged (or unless they were pushed by accident). The whole point of version management is to keep track of history. When we start deleting unmerged branches, we lose history forever.

We had this discussion before btw: https://github.com/NixOS/nixpkgs/issues/12288

**Comment by shlevy** [comment]
I disagree that we actually do lose history forever, as everyone who cloned before the deletion has it and can restore it at any point. What's the point of dvcs if not this? I also disagree that all history is worth keeping track of, *especially* on a shared mainline development tree. Would you propose we add branch protection to *all* branches so we can never force-push either?

There is a meaningful distinction between "this is a permanent reference with reliable history" and "this is a shared community effort that may or may not go anywhere and may be changed at any time". Git branches on the mainline tree are a perfectly reasonable way to express both, provided we layer a policy on top (e.g. branch protection, a list of reliable branches, etc.).

**Comment by domenkozar** [comment]
I share concerns with @edolstra here, what's the benefit of this OCD?

**Comment by lheckemann** [comment]
>what's the benefit of this OCD?

As shlevy wrote,

>they litter git branch -a and the github branch list

I frequently use `git branch -a`, but it's not really useful for nixpkgs because there are so many branches I don't care about.

**Comment by shlevy** [comment]
It also ruins tab-completion when checking out a branch tracked from upstream for the first time.

There's really no call to call this request obsessive-compulsive or disordered, even if you disagree...

**Comment by shlevy** [comment]
Anyway, if we're actually going to take a hard line "history is history and can't ever be deleted" stance, then we should disable all force-pushes and probably have a stricter approval process for creating new branches.

**Comment by copumpkin** [comment]
Agree, it just looks messy and offers little benefit. If the branch names were somehow annotated with their purpose, it might be an interesting historical artifact, but even then I'd probably keep the pointers somewhere else.

**Comment by edolstra** [comment]
If we start deleting branches, then our Git repo cannot be relied on anymore as a place where we can safely keep our sources. It means we should warn users that e.g. `fetchFromGitHub` might not work in the future because referenced revisions might disappear. Also, will there be a backup? What if we delete a branch and two years later realize it may have contained something useful? It all seems pretty risky just to make tab completion work better.

@shlevy Sure, force-pushes should be disabled but I don't think GitHub allows that for an entire repository. I also agree that creating new branches should be limited (e.g. only to collaborations between multiple people).

**Comment by edolstra** [comment]
Just to give an example of why deleting history is bad: some of the branches that are proposed to be deleted were used to produce the results for published papers (e.g. `purity-test` is referenced in the ICFP-2008 paper on NixOS, and I believe `syscall-tracing` was used for the ASE-2014 paper on build process tracing). So from the perspective of reproducible research, creating a new policy of deleting old data is a bad idea.

**Comment by bjornfor** [comment]
> If we start deleting branches, then our Git repo cannot be relied on anymore as a place where we can safely keep our sources. It means we should warn users that e.g. fetchFromGitHub might not work in the future because referenced revisions might disappear.

I personally would never rely on any nixpkgs branch to contain stable history except for a few select branches: master, release-*. Anything happening on other branches is not part of nixpkgs history as far as I'm concerned. Things not reachable from current, active branches "didn't happen".

RE: branches referenced by papers. Those branches could be replaced by tags. But I guess in this case it only shifts the problem from branches to tags. But IMHO, it's still better. Branches are expected to move, tags not. Stale branches provide little value. Also, for reproducibility you want a fixed point in history (tag) not a floating one (branch).

**Comment by bjornfor** [comment]
(Arguably, the branches referenced by papers are special cases.)

**Comment by shlevy** [comment]
Branches referenced by papers make sense to be in the "stable" set, though I agree for a specific publication a tag makes more sense.

For now, if we can't agree on deletion what I'd like to do is prohibit all new branches except release branches and master, put branch protection on all branches, and make a new repo for the ephemeral work. I'd ideally like to move all the current "stale" branches somewhere outside of the main repo, but I understand if retroactive moves are a problem due to link stability.

Who has admin access to the repo to turn on protection on the branches we have?

**Comment by davidak** [comment]
I don't want to delete valuable work and don't want to break links.

Can anyone delete the merged branches?

Can we also delete `meteor` and `nix-upgrade` since it's implemented by different PR?

Then we can close this issue and move on to more important stuff. Thanks for the discussion.

**Comment by fpletz** [comment]
`staging-17.03` is definitely not needed anymore and was merged into `release-17.03`. I've removed it.

**Comment by c0bw3b** [comment]
Deleted `meteor` since all its history is already in master 25e582c497699b3a3b8aff8552d902933986ae05

No consensus on the rest so closing.

**Comment by Profpatsch** [comment]
Have all merged branches been deleted already?

**Comment by lheckemann** [comment]
This also contributes to triggering of https://github.com/NixOS/nix/issues/2147 in commitIdFromGitRepo since all the remote refs are included in .git/packed-refs, which is processed using `builtins.match`.
